File: src/test/java/com/handwoong/everyonewaiter/service/MemberServiceImplTest.java
package com.handwoong.everyonewaiter.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.handwoong.everyonewaiter.domain.Member;
import com.handwoong.everyonewaiter.dto.MemberDto;
import com.handwoong.everyonewaiter.dto.OnlyMsgResDto;
import com.handwoong.everyonewaiter.enums.MemberRole;
import com.handwoong.everyonewaiter.exception.CustomException;
import com.handwoong.everyonewaiter.repository.FakeMemberRepository;
import java.util.List;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.security.crypto.password.PasswordEncoder;

class MemberServiceImplTest {

    private MemberServiceImpl memberService;

    private FakeMemberRepository memberRepository;

    private PasswordEncoder passwordEncoder;

    private MemberDto.RequestDto memberDto;

    @BeforeEach
    void beforeEach() {
        passwordEncoder = Mockito.mock(PasswordEncoder.class);
        memberRepository = new FakeMemberRepository();
        memberService = new MemberServiceImpl(memberRepository, passwordEncoder);
        memberDto = new MemberDto.RequestDto("handwoong", "password1", "01012345678");
    }

    @AfterEach
    void afterEach() {
        memberRepository.clear();
    }

    @Test
    @DisplayName("회원가입 성공")
    void register() throws Exception {
        // given
        when(passwordEncoder.encode(anyString())).thenReturn("encodedPassword");

        // when
        memberService.register(memberDto);
        Member member = memberRepository.findById(1L).orElseThrow();

        // then
        verify(passwordEncoder, times(1)).encode(memberDto.getPassword());
        assertThat(member.getUsername()).isEqualTo(memberDto.getUsername());
        assertThat(member.getPassword()).isEqualTo("encodedPassword");
        assertThat(member.getPhoneNumber()).isEqualTo(memberDto.getPhoneNumber());
        assertThat(member.getRole()).isEqualTo(MemberRole.ROLE_USER);
        assertThat(member.getBalance()).isEqualTo(300);
    }

    @Test
    @DisplayName("회원가입 시 로그인 아이디 중복 예외")
    void registerExistsUsername() throws Exception {
        // given
        memberService.register(memberDto);

        // when
        assertThatThrownBy(() -> memberService.register(memberDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
    }

    @Test
    @DisplayName("회원가입 시 휴대폰 번호 중복 예외")
    void registerExistsPhoneNumber() throws Exception {
        // given
        memberService.register(memberDto);
        memberDto.setUsername("username");

        // when
        assertThatThrownBy(() -> memberService.register(memberDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
    }

    @Test
    @DisplayName("회원 ID 조회")
    void findMemberById() throws Exception {
        // given
        memberService.register(memberDto);

        // when
        MemberDto.ResponseDto member = memberService.findMember(1L);

        // then
        assertThat(member.getUsername()).isEqualTo(memberDto.getUsername());
        assertThat(member.getPhoneNumber()).isEqualTo(memberDto.getPhoneNumber());
        assertThat(member.getBalance()).isEqualTo(300);
    }

    @Test
    @DisplayName("회원 존재하지 않는 ID 조회")
    void findNotExistsById() throws Exception {
        // given
        assertThatThrownBy(() -> memberService.findMember(1L))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
    }

    @Test
    @DisplayName("회원 Username 조회")
    void findMemberByUsername() throws Exception {
        // given
        memberService.register(memberDto);

        // when
        MemberDto.ResponseDto member = memberService.findMemberByUsername("handwoong");

        // then
        assertThat(member.getUsername()).isEqualTo(memberDto.getUsername());
        assertThat(member.getPhoneNumber()).isEqualTo(memberDto.getPhoneNumber());
        assertThat(member.getBalance()).isEqualTo(300);
    }

    @Test
    @DisplayName("회원 존재하지 않는 Username 조회")
    void findNotExistsByUsername() throws Exception {
        assertThatThrownBy(() -> memberService.findMemberByUsername("handwoong"))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
    }

    @Test
    @DisplayName("회원 목록 조회")
    void findMemberList() throws Exception {
        // given
        memberService.register(memberDto);
        memberDto.setUsername("handwoong@test.com");
        memberDto.setPhoneNumber("01011112222");
        memberService.register(memberDto);

        // when
        List<MemberDto.ResponseDto> memberList = memberService.findMemberList();
        MemberDto.ResponseDto memberDtoA = memberList.get(0);
        MemberDto.ResponseDto memberDtoB = memberList.get(1);

        // then
        assertThat(memberList.size()).isEqualTo(2);
        assertThat(memberDtoA.getUsername()).isEqualTo("handwoong");
        assertThat(memberDtoB.getUsername()).isEqualTo("handwoong@test.com");
        assertThat(memberDtoA.getPhoneNumber()).isEqualTo("01012345678");
        assertThat(memberDtoB.getPhoneNumber()).isEqualTo("01011112222");
    }

    @Test
    @DisplayName("회원 비밀번호 변경")
    void changeMemberPassword() throws Exception {
        // given
        memberDto.setPassword("encodedPassword");
        Member member = Member.createMember(memberDto);
        memberRepository.save(member);
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto(member.getPassword(), "newPassword");

        when(passwordEncoder.matches(anyString(), anyString())).thenReturn(true);
        when(passwordEncoder.encode(anyString())).thenReturn("encodedPassword");

        // when
        OnlyMsgResDto result = memberService
                .changePassword(memberDto.getUsername(), passwordDto);

        // then
        assertThat(result.getMessage()).isEqualTo("success");
        verify(passwordEncoder, times(1)).encode("newPassword");
        verify(passwordEncoder, times(1)).matches("encodedPassword", "encodedPassword");
    }

    @Test
    @DisplayName("회원 비밀번호 변경 시 기존 비밀번호가 일치하지 않으면 예외 발생")
    void notMatchCurrentPassword() throws Exception {
        // given
        memberDto.setPassword("encodedPassword");
        Member member = Member.createMember(memberDto);
        memberRepository.save(member);
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto("diffPassword", "newPassword");

        assertThatThrownBy(
                () -> memberService.changePassword(member.getUsername(), passwordDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
        verify(passwordEncoder, times(0)).encode(anyString());
    }

    @Test
    @DisplayName("회원 비밀번호 변경 시 회원을 찾을 수 없으면 예외 발생")
    void notFoundMember() throws Exception {
        // given
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto("encodedPassword", "newPassword");

        assertThatThrownBy(
                () -> memberService.changePassword("handwoong", passwordDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
        verify(passwordEncoder, times(0)).encode(anyString());
        verify(passwordEncoder, times(0)).matches(anyString(), anyString());
    }

    @Test
    @DisplayName("회원 탈퇴")
    void deleteMember() throws Exception {
        // given
        Member member = Member.createMember(memberDto);
        memberRepository.save(member);
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto(member.getPassword(), "newPassword");
        when(passwordEncoder.matches(anyString(), anyString())).thenReturn(true);

        // when
        OnlyMsgResDto result = memberService.deleteMember(member.getUsername(), passwordDto);
        long count = memberRepository.count();

        // then
        assertThat(result.getMessage()).isEqualTo("success");
        assertThat(count).isEqualTo(0);
    }

    @Test
    @DisplayName("회원 탈퇴 시 존재하지 않는 아이디면 예외 발생")
    void deleteMemberNotFound() throws Exception {
        // given
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto(memberDto.getPassword(), "newPassword");

        // when
        assertThatThrownBy(
                () -> memberService.deleteMember(memberDto.getUsername(), passwordDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");

        // then
        verify(passwordEncoder, times(0)).matches(anyString(), anyString());
    }

    @Test
    @DisplayName("회원 탈퇴 시 비밀번호가 일치하지 않으면 예외 발생")
    void deleteMemberNotMatchPassword() throws Exception {
        // given
        Member member = Member.createMember(memberDto);
        memberRepository.save(member);
        MemberDto.PwdRequestDto passwordDto =
                new MemberDto.PwdRequestDto(memberDto.getPassword(), "newPassword");

        // when
        assertThatThrownBy(
                () -> memberService.deleteMember(memberDto.getUsername(), passwordDto))
                .isInstanceOf(CustomException.class)
                .hasOnlyFields("errorCode");
        long count = memberRepository.count();

        // then
        verify(passwordEncoder, times(1)).matches(anyString(), anyString());
        assertThat(count).isEqualTo(1);
    }
}

File: src/test/kotlin/com/handwoong/everyonewaiter/service/member/MemberServiceImplTest.kt
package com.handwoong.everyonewaiter.service.member

import com.handwoong.everyonewaiter.domain.member.Member
import com.handwoong.everyonewaiter.domain.member.MemberRole
import com.handwoong.everyonewaiter.dto.member.MemberCreateRequest
import com.handwoong.everyonewaiter.dto.member.PasswordRequest
import com.handwoong.everyonewaiter.exception.CustomException
import com.handwoong.everyonewaiter.exception.ErrorCode.*
import com.handwoong.everyonewaiter.repository.member.MemberRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.transaction.annotation.Transactional

@Transactional
@SpringBootTest
class MemberServiceImplTest @Autowired constructor(
    private val memberService: MemberService,
    private val memberRepository: MemberRepository,
    private val passwordEncoder: PasswordEncoder,
) {

    @Test
    @DisplayName("회원가입에 성공한다.")
    fun registerSuccessTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()

        // when
        memberService.register(memberCreateRequest)

        // then
        val memberList = memberRepository.findAll()
        assertThat(memberList).hasSize(1)
        assertThat(memberList[0].username).isEqualTo("username")
        assertThat(memberList[0].phoneNumber).isEqualTo("01012345678")
        assertThat(memberList[0].role).isEqualTo(MemberRole.ROLE_USER)
    }

    @Test
    @DisplayName("회원가입 시 로그인 아이디(username)가 이미 존재하면 예외가 발생한다.")
    fun registerExistsUsernameTest() {
        // given
        val memberCreateRequest1 = MemberCreateRequest.testOf()
        val memberCreateRequest2 = memberCreateRequest1.copy(phoneNumber = "01011112222")
        memberService.register(memberCreateRequest1)

        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.register(memberCreateRequest2)
        }.errorCode
        assertThat(errorCode).isEqualTo(USERNAME_OR_PHONE_EXISTS)
    }

    @Test
    @DisplayName("회원가입 시 휴대폰 번호(phoneNumber)가 이미 존재하면 예외가 발생한다.")
    fun registerExistsPhoneNumberTest() {
        // given
        val request1 = MemberCreateRequest.testOf()
        val request2 = request1.copy(username = "username")
        memberService.register(request1)

        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.register(request2)
        }.errorCode
        assertThat(errorCode).isEqualTo(USERNAME_OR_PHONE_EXISTS)
    }

    @Test
    @DisplayName("회원가입 시 비밀번호 암호화가 정상적으로 이루어진다.")
    fun registerEncodePasswordTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()

        // when
        memberService.register(memberCreateRequest)

        // then
        val memberList = memberRepository.findAll()
        assertThat(memberList[0].password).isNotEqualTo(memberCreateRequest.password)
        assertThat(memberList[0].password).contains("bcrypt")
    }

    @Test
    @DisplayName("회원 ID로 회원을 조회한다.")
    fun findMemberByIdTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()
        val member = Member.createMember(memberCreateRequest)
        val saveMember = memberRepository.save(member)

        // when
        val findMember = memberService.findMemberById(saveMember.id!!)

        // then
        assertThat(findMember.id).isEqualTo(saveMember.id)
        assertThat(findMember.username).isEqualTo("username")
        assertThat(findMember.phoneNumber).isEqualTo("01012345678")
        assertThat(findMember.balance).isEqualTo(300)
    }

    @Test
    @DisplayName("회원 ID로 회원 조회 시 회원이 존재하지 않으면 예외가 발생한다.")
    fun findMemberByIdNoSuchMemberTest() {
        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.findMemberById(1L)
        }.errorCode
        assertThat(errorCode).isEqualTo(ID_RESOURCE_NOT_FOUND)
    }

    @Test
    @DisplayName("회원 로그인 아이디(username)로 회원을 조회한다.")
    fun findMemberByUsernameTest() {
        // given
        val member = Member.createMember(MemberCreateRequest.testOf())
        val saveMember = memberRepository.save(member)

        // when
        val findMember = memberService.findMemberByUsername("username")

        // then
        assertThat(findMember.id).isEqualTo(saveMember.id)
        assertThat(findMember.username).isEqualTo("username")
        assertThat(findMember.phoneNumber).isEqualTo("01012345678")
        assertThat(findMember.balance).isEqualTo(300)
    }

    @Test
    @DisplayName("회원 로그인 아이디(username)로 조회 시 회원이 존재하지 않으면 예외가 발생한다.")
    fun findMemberByUsernameNoSuchMemberTest() {
        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.findMemberByUsername("username")
        }.errorCode
        assertThat(errorCode).isEqualTo(MEMBER_NOT_FOUND)
    }

    @Test
    @DisplayName("모든 회원 목록을 조회한다.")
    fun findAllMemberListTest() {
        // given
        memberRepository.saveAll(
            listOf(
                Member.createMember(MemberCreateRequest.testOf()),
                Member.createMember(MemberCreateRequest.testOf(username = "handwoong", phoneNumber = "01011112222")),
            )
        )

        // when
        val memberList = memberService.findAllMemberList()

        // then
        assertThat(memberList).hasSize(2)
        assertThat(memberList[0].username).isEqualTo("username")
        assertThat(memberList[1].username).isEqualTo("handwoong")
        assertThat(memberList[0].phoneNumber).isEqualTo("01012345678")
        assertThat(memberList[1].phoneNumber).isEqualTo("01011112222")
    }

    @Test
    @DisplayName("모든 회원 목록 조회 시 회원이 한명도 없다면 빈 리스트를 반환한다.")
    fun findAllMemberListNoSuchMemberTest() {
        // when
        val memberList = memberService.findAllMemberList()

        // then
        assertThat(memberList).isEmpty()
    }

    @Test
    @DisplayName("회원 비밀번호가 성공적으로 변경된다.")
    fun changePasswordSuccessTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()
        memberService.register(memberCreateRequest)
        val passwordRequest = PasswordRequest(
            currentPassword = "password1",
            newPassword = "newPassword1",
        )

        // when
        memberService.changePassword("username", passwordRequest)

        // then
        val memberList = memberRepository.findAll()
        assertThat(memberList[0].password).contains("bcrypt")
        assertThat(
            passwordEncoder.matches(
                passwordRequest.newPassword,
                memberList[0].password,
            )
        ).isTrue()
    }

    @Test
    @DisplayName("회원 비밀번호 변경 시 회원을 찾지 못하면 예외가 발생한다.")
    fun changePasswordNoSuchMemberTest() {
        // given
        val passwordRequest = PasswordRequest(
            currentPassword = "password1",
            newPassword = "newPassword1",
        )

        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.changePassword("username", passwordRequest)
        }.errorCode
        assertThat(errorCode).isEqualTo(MEMBER_NOT_FOUND)
    }

    @Test
    @DisplayName("회원 비밀번호 변경 시 현재 비밀번호와 저장된 비밀번호가 다르다면 예외가 발생한다.")
    fun changePasswordNotMatchPasswordTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()
        memberService.register(memberCreateRequest)
        val passwordRequest = PasswordRequest(
            // password1 != password2
            currentPassword = "password2",
            newPassword = "newPassword1",
        )

        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.changePassword("username", passwordRequest)
        }.errorCode
        assertThat(errorCode).isEqualTo(NOT_MATCH_PASSWORD)
    }

    @Test
    @DisplayName("회원 삭제가 정상 동작한다.")
    fun deleteMemberTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()
        memberService.register(memberCreateRequest)
        val passwordRequest = PasswordRequest(
            currentPassword = "password1",
            newPassword = "newPassword1",
        )

        // when
        memberService.deleteMember("username", passwordRequest)

        // then
        val memberList = memberRepository.findAll()
        assertThat(memberList).isEmpty()
    }

    @Test
    @DisplayName("회원 삭제 시 현재 비밀번호와 저장된 비밀번호가 다르다면 예외가 발생한다.")
    fun deleteMemberNotMatchPasswordTest() {
        // given
        val memberCreateRequest = MemberCreateRequest.testOf()
        memberService.register(memberCreateRequest)
        val passwordRequest = PasswordRequest(
            // password1 != password2
            currentPassword = "password2",
            newPassword = "newPassword1",
        )

        // when && then
        val errorCode = assertThrows<CustomException> {
            memberService.deleteMember("username", passwordRequest)
        }.errorCode
        assertThat(errorCode).isEqualTo(NOT_MATCH_PASSWORD)
    }

}
